{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bf3947ac-7539-4918-ab04-ab02647bc010",
   "metadata": {},
   "source": [
    "# IVF Study #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cd26cd0c-ac80-4450-a47a-f546a3cc16d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "go-basic.obo: fmt(1.2) rel(2025-07-22) 43,230 Terms\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'term' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 12\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mgoatools\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mobo_parser\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m GODag\n\u001b[1;32m     11\u001b[0m go_dag \u001b[38;5;241m=\u001b[39m GODag(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgo-basic.obo\u001b[39m\u001b[38;5;124m\"\u001b[39m, optional_attrs\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdefinition\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[0;32m---> 12\u001b[0m definition \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(term, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdefinition\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;28mgetattr\u001b[39m(term, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdefn\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m))\u001b[38;5;241m.\u001b[39mlower()\n",
      "\u001b[0;31mNameError\u001b[0m: name 'term' is not defined"
     ]
    }
   ],
   "source": [
    "from sqlalchemy import create_engine\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "import urllib.request\n",
    "url = \"http://purl.obolibrary.org/obo/go/go-basic.obo\"\n",
    "urllib.request.urlretrieve(url, \"go-basic.obo\")\n",
    "from goatools.obo_parser import GODag\n",
    "go_dag = GODag(\"go-basic.obo\", optional_attrs=['definition'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b059625-2764-4b18-868d-c1dc2b25361e",
   "metadata": {},
   "source": [
    "### Connect to the SQL database. ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b02fbada-b06b-44a4-a5d1-548b47882594",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PostgreSQL connection details\n",
    "host = \"localhost\"\n",
    "dbname = \"IVF\"\n",
    "user = \"postgres\"\n",
    "password = os.getenv(\"PASSWORD\")\n",
    "\n",
    "# Create an SQLAlchemy engine\n",
    "engine = create_engine(f\"postgresql://{user}:{password}@{host}/{dbname}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a11b7907-0d73-4291-b103-b6c6ccaa03bf",
   "metadata": {},
   "source": [
    "### We begin with a data summary. ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "811bf17f-b3e9-4dcb-98ff-9966eb2dcb63",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize the summary data structure\n",
    "summary_df = pd.DataFrame(columns=[\n",
    "    'patient_id',\n",
    "\n",
    "    # 2a: Variant Burden & Frequency\n",
    "    'count_baseline_variants',\n",
    "    'count_somatic_variants',\n",
    "    'mean_af_baseline',\n",
    "    'mean_af_somatic',\n",
    "    'count_af_over_20_baseline',\n",
    "    'count_af_over_20_somatic',\n",
    "    'shared_loci_count',\n",
    "\n",
    "    # 2b: Functional Impact\n",
    "    'pathogenic_clinvar_count',\n",
    "    'high_impact_function_count',\n",
    "    'damaging_sift_count',\n",
    "    'damaging_polyphen_count',\n",
    "    'mean_phred_qual_score',\n",
    "\n",
    "    # 2c: Gene & Pathway Features\n",
    "    'ivf_gene_variant_count',\n",
    "    'reproductive_go_variant_count',\n",
    "    'unique_genes_affected'\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e94007a5-7a29-4e45-b61f-bdf5120002c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_ids = [1, 2, 3, 4, 5, 6]\n",
    "results = []\n",
    "\n",
    "for pid in patient_ids:\n",
    "    \n",
    "    # Count Variants\n",
    "    count_baseline_query = f\"\"\" \n",
    "        SELECT COUNT(*) AS count \n",
    "        FROM all_baseline \n",
    "        WHERE patient_id = {pid} \n",
    "    \"\"\"\n",
    "    count_baseline = pd.read_sql(count_baseline_query, engine).iloc[0]['count']\n",
    "    \n",
    "    count_somatic_query = f\"\"\" \n",
    "        SELECT COUNT(*) AS count \n",
    "        FROM all_somatic \n",
    "        WHERE patient_id = {pid} \n",
    "    \"\"\"\n",
    "    count_somatic = pd.read_sql(count_somatic_query, engine).iloc[0]['count']\n",
    "\n",
    "    # Allele frequency\n",
    "    allele_freq_baseline_query = f\"\"\" \n",
    "        SELECT allele_frequency_percent \n",
    "        FROM all_baseline \n",
    "        WHERE patient_id = {pid} \n",
    "    \"\"\"\n",
    "    allele_freq_baseline_df = pd.read_sql(allele_freq_baseline_query, engine)\n",
    "    allele_freq_baseline_df['source'] = 'baseline'\n",
    "\n",
    "    allele_freq_somatic_query = f\"\"\" \n",
    "        SELECT allele_frequency_percent \n",
    "        FROM all_somatic \n",
    "        WHERE patient_id = {pid} \n",
    "    \"\"\"\n",
    "    allele_freq_somatic_df = pd.read_sql(allele_freq_somatic_query, engine)\n",
    "    allele_freq_somatic_df['source'] = 'somatic'\n",
    "    \n",
    "    # Combine into one DataFrame\n",
    "    allele_freq_df = pd.concat([allele_freq_baseline_df, allele_freq_somatic_df], ignore_index=True)\n",
    "    allele_freq_df['allele_frequency_percent'] = pd.to_numeric(allele_freq_df['allele_frequency_percent'], errors='coerce')\n",
    "\n",
    "    # Compute the mean allele frequencies\n",
    "    mean_af_source = allele_freq_df.groupby('source')['allele_frequency_percent'].mean()\n",
    "    mean_af_baseline = mean_af_source.get('baseline', float('nan'))\n",
    "    mean_af_somatic = mean_af_source.get('somatic', float('nan'))\n",
    "\n",
    "    # Counts over threshold\n",
    "    allele_freq_df['af_gt_20'] = allele_freq_df['allele_frequency_percent'] > 20\n",
    "    count_af_over_20_source = allele_freq_df.groupby('source')['af_gt_20'].sum()\n",
    "    count_af_over_20_baseline = count_af_over_20_source.get('baseline', float('nan'))\n",
    "    count_af_over_20_somatic = count_af_over_20_source.get('somatic', float('nan'))\n",
    "\n",
    "    # Shared loci count\n",
    "    baseline_loci_query = f\"\"\"\n",
    "        SELECT DISTINCT locus\n",
    "        FROM all_baseline\n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    baseline_loci_df = pd.read_sql(baseline_loci_query, engine)\n",
    "    baseline_loci = set(baseline_loci_df['locus'])\n",
    "\n",
    "    somatic_loci_query = f\"\"\"\n",
    "        SELECT DISTINCT locus\n",
    "        FROM all_somatic\n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    somatic_loci_df = pd.read_sql(somatic_loci_query, engine)\n",
    "    somatic_loci = set(somatic_loci_df['locus'])\n",
    "\n",
    "    # Find shared loci\n",
    "    shared_loci = baseline_loci.intersection(somatic_loci)\n",
    "    shared_loci_count = len(shared_loci)\n",
    "\n",
    "    # Pathogenic count  \n",
    "    clinvar_baseline_query = f\"\"\" \n",
    "        SELECT clinvar \n",
    "        FROM all_baseline \n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    clinvar_baseline_df = pd.read_sql(clinvar_baseline_query, engine)\n",
    "    clinvar_baseline_df['source'] = 'baseline'\n",
    "    clinvar_somatic_query = f\"\"\" \n",
    "        SELECT clinvar \n",
    "        FROM all_somatic \n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    clinvar_somatic_df = pd.read_sql(clinvar_somatic_query, engine)\n",
    "    clinvar_somatic_df['source'] = 'somatic'\n",
    "    clinvar_df = pd.concat([clinvar_baseline_df, clinvar_somatic_df], ignore_index=True)\n",
    "    pathogenic_count = clinvar_df[clinvar_df['clinvar'].str.contains('pathogenic', case=False, na=False)]['clinvar'].count()\n",
    "\n",
    "    # High impact function count\n",
    "    high_impact_terms = ['frameshift', 'stop_gained', 'splice_donor', 'splice_acceptor']\n",
    "    function_baseline_query = f\"\"\" \n",
    "        SELECT function \n",
    "        FROM all_baseline \n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    function_baseline_df = pd.read_sql(function_baseline_query, engine)\n",
    "    function_baseline_df['source'] = 'baseline'\n",
    "    function_somatic_query = f\"\"\" \n",
    "        SELECT function \n",
    "        FROM all_somatic \n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    function_somatic_df = pd.read_sql(function_somatic_query, engine)\n",
    "    function_somatic_df['source'] = 'somatic'\n",
    "    function_df = pd.concat([function_baseline_df, function_somatic_df], ignore_index=True)\n",
    "    high_impact_function_count = function_df[function_df['function'].str.contains('|'.join(high_impact_terms), case=False, na=False)]['function'].count()\n",
    "\n",
    "    # Damaging sift count\n",
    "    sift_baseline_query = f\"\"\" \n",
    "        SELECT sift \n",
    "        FROM all_baseline \n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    sift_baseline_df = pd.read_sql(sift_baseline_query, engine)\n",
    "    sift_baseline_df['source'] = 'baseline'\n",
    "    sift_somatic_query = f\"\"\" \n",
    "        SELECT sift \n",
    "        FROM all_somatic \n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    sift_somatic_df = pd.read_sql(sift_somatic_query, engine)\n",
    "    sift_somatic_df['source'] = 'somatic'\n",
    "    sift_df = pd.concat([sift_baseline_df, sift_somatic_df], ignore_index=True)\n",
    "    sift_df['sift'] = pd.to_numeric(sift_df['sift'], errors='coerce')\n",
    "    damaging_sift_count = sift_df[sift_df['sift'] < 0.05]['sift'].count()\n",
    "\n",
    "    # Damaging polyphen count\n",
    "    polyphen_baseline_query = f\"\"\" \n",
    "        SELECT polyphen \n",
    "        FROM all_baseline \n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    polyphen_baseline_df = pd.read_sql(polyphen_baseline_query, engine)\n",
    "    polyphen_baseline_df['source'] = 'baseline'\n",
    "    polyphen_somatic_query = f\"\"\" \n",
    "        SELECT polyphen \n",
    "        FROM all_somatic \n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    polyphen_somatic_df = pd.read_sql(polyphen_somatic_query, engine)\n",
    "    polyphen_somatic_df['source'] = 'somatic'\n",
    "    polyphen_df = pd.concat([polyphen_baseline_df, polyphen_somatic_df], ignore_index=True)\n",
    "    polyphen_df['polyphen'] = pd.to_numeric(polyphen_df['polyphen'], errors='coerce')\n",
    "    damaging_polyphen_count = polyphen_df[polyphen_df['polyphen'] > 0.85]['polyphen'].count()\n",
    "\n",
    "    # Mean phred qual score\n",
    "    phred_qual_score_baseline_query = f\"\"\" \n",
    "        SELECT phred_qual_score \n",
    "        FROM all_baseline \n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    phred_qual_score_baseline_df = pd.read_sql(phred_qual_score_baseline_query, engine)\n",
    "    phred_qual_score_baseline_df['source'] = 'baseline'\n",
    "    phred_qual_score_somatic_query = f\"\"\" \n",
    "        SELECT phred_qual_score \n",
    "        FROM all_somatic \n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    phred_qual_score_somatic_df = pd.read_sql(phred_qual_score_somatic_query, engine)\n",
    "    phred_qual_score_somatic_df['source'] = 'somatic'\n",
    "    phred_qual_score_df = pd.concat([phred_qual_score_baseline_df, phred_qual_score_somatic_df], ignore_index=True)\n",
    "    phred_qual_score_df['phred_qual_score'] = pd.to_numeric(phred_qual_score_df['phred_qual_score'], errors='coerce')\n",
    "    mean_phred_qual_score = phred_qual_score_df['phred_qual_score'].mean()\n",
    "\n",
    "    # IVF gene variant count\n",
    "    ivf_genes = ['BRCA1', 'BRCA2', 'FMR1', 'NR5A1', 'ESR1', 'FSHR', 'LHCGR', 'AMH', 'BMP15', 'GDF9', 'ZP3', 'ZP1', 'ZP2', 'NANOS3', 'NOBOX', 'FIGLA']\n",
    "    gene_baseline_query = f\"\"\" \n",
    "        SELECT gene\n",
    "        FROM all_baseline \n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    gene_baseline_df = pd.read_sql(gene_baseline_query, engine)\n",
    "    gene_baseline_df['source'] = 'baseline'\n",
    "    gene_somatic_query = f\"\"\" \n",
    "        SELECT gene \n",
    "        FROM all_somatic \n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    gene_somatic_df = pd.read_sql(gene_somatic_query, engine)\n",
    "    gene_somatic_df['source'] = 'somatic'\n",
    "    gene_df = pd.concat([gene_baseline_df, gene_somatic_df], ignore_index=True)\n",
    "    ivf_gene_variant_count = gene_df[gene_df['gene'].isin(ivf_genes)]['gene'].count()\n",
    "    unique_genes_affected = gene_df['gene'].nunique()\n",
    "    \n",
    "    # Reproductive go variant count\n",
    "    repro_terms = ['reproduction', 'fertilization', 'oogenesis', 'spermatogenesis', 'embryo','gonad', 'infertility', 'implantation', 'oocyte', 'sperm', 'zygote','follicle', 'blastocyst', 'luteinizing', 'FSH', 'hormone', 'gonadotropin']\n",
    "    repro_go_ids = set()\n",
    "    for go_id, term in go_dag.items():\n",
    "        definition = getattr(term, 'definition', getattr(term, 'defn', '')).lower()\n",
    "        name = term.name.lower()\n",
    "        definition = term.definition.lower() if term.definition else ''\n",
    "        if any(r in name or r in definition for r in repro_terms):\n",
    "            repro_go_ids.add(go_id)\n",
    "    go_baseline_query = f\"\"\" \n",
    "        SELECT go\n",
    "        FROM all_baseline \n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    go_baseline_df = pd.read_sql(go_baseline_query, engine)\n",
    "    go_baseline_df['source'] = 'baseline'\n",
    "    go_somatic_query = f\"\"\" \n",
    "        SELECT go \n",
    "        FROM all_somatic \n",
    "        WHERE patient_id = {pid}\n",
    "    \"\"\"\n",
    "    go_somatic_df = pd.read_sql(go_somatic_query, engine)\n",
    "    go_somatic_df['source'] = 'somatic'\n",
    "    go_df = pd.concat([go_baseline_df, go_somatic_df], ignore_index=True)\n",
    "    reproductive_go_variant_count = go_df[go_df['go'].str.contains('|'.join(repro_go_ids), case=False, na=False)]['go'].count()\n",
    "\n",
    "    # Append results\n",
    "    results.append({\n",
    "        'patient_id': pid,\n",
    "        'count_baseline_variants': count_baseline,\n",
    "        'count_somatic_variants': count_somatic,\n",
    "        'mean_af_baseline': mean_af_baseline,\n",
    "        'mean_af_somatic': mean_af_somatic,\n",
    "        'count_af_over_20_baseline': count_af_over_20_baseline,\n",
    "        'count_af_over_20_somatic': count_af_over_20_somatic,\n",
    "        'shared_loci_count': shared_loci_count,\n",
    "        'pathogenic_clinvar_count': pathogenic_count,\n",
    "        'high_impact_function_count': high_impact_function_count,\n",
    "        'damaging_sift_count': damaging_sift_count,\n",
    "        'damaging_polyphen_count': damaging_polyphen_count,\n",
    "        'mean_phred_qual_score': mean_phred_qual_score,\n",
    "        'ivf_gene_variant_count': ivf_gene_variant_count,\n",
    "        'reproductive_go_variant_count': reproductive_go_variant_count,\n",
    "        'unique_genes_affected': unique_genes_affected\n",
    "    })\n",
    "\n",
    "# Create summary DataFrame\n",
    "summary_df = pd.DataFrame(results)\n",
    "print(summary_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e896719-b2bc-4088-9816-651811027c95",
   "metadata": {},
   "outputs": [],
   "source": [
    "sift_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70865cbc-56b2-4cbc-be99-6e5d75722898",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Query to get allele frequency for both somatic and non-somatic data\n",
    "query = \"\"\"\n",
    "    SELECT patient, allele_frequency, somatic_allele_frequency \n",
    "    FROM allele_frequency_updated\n",
    "    WHERE allele_frequency IS NOT NULL AND somatic_allele_frequency IS NOT NULL;\n",
    "\"\"\"\n",
    "\n",
    "# Read data into a pandas DataFrame\n",
    "df = pd.read_sql(query, engine)\n",
    "\n",
    "# Group data by patient\n",
    "patients = df['patient'].unique()\n",
    "\n",
    "# Plot for each patient\n",
    "for patient in patients:\n",
    "    patient_data = df[df['patient'] == patient]\n",
    "    \n",
    "    # Scatter plot for somatic vs non-somatic allele frequency\n",
    "    plt.scatter(patient_data['allele_frequency'], patient_data['somatic_allele_frequency'], \n",
    "                label=f'Patient {patient}', alpha=0.5)\n",
    "    plt.xlabel('allele frequency (%)')\n",
    "    plt.ylabel('Somatic allele frequency (%)')\n",
    "    plt.title(f'Patient {patient}: Somatic vs Non-somatic Allele Frequencies')\n",
    "    plt.legend()\n",
    "    plt.grid(True)\n",
    "    plt.show()\n",
    "\n",
    "# Optional: Combine all patients into a single plot\n",
    "# plt.scatter(df['non_somatic_allele_freq'], df['somatic_allele_frequency_percent'], alpha=0.5)\n",
    "# plt.xlabel('Non-somatic allele frequency (%)')\n",
    "# plt.ylabel('Somatic allele frequency (%)')\n",
    "# plt.title('Somatic vs Non-somatic Allele Frequencies (All Patients)')\n",
    "# plt.grid(True)\n",
    "# plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42c9aa4c-6974-4e37-89f1-fb5231333bd8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5598472-3086-4dcf-b9ed-eab54445a97c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
