{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bf3947ac-7539-4918-ab04-ab02647bc010",
   "metadata": {},
   "source": [
    "# IVF Study #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cd26cd0c-ac80-4450-a47a-f546a3cc16d0",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "go-basic.obo: fmt(1.2) rel(2025-07-22) 43,230 Terms\n"
     ]
    }
   ],
   "source": [
    "from sqlalchemy import create_engine\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "import urllib.request\n",
    "url = \"http://purl.obolibrary.org/obo/go/go-basic.obo\"\n",
    "urllib.request.urlretrieve(url, \"go-basic.obo\")\n",
    "from goatools.obo_parser import GODag\n",
    "go_dag = GODag(\"go-basic.obo\", optional_attrs=['definition'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b059625-2764-4b18-868d-c1dc2b25361e",
   "metadata": {},
   "source": [
    "### Connect to the SQL database. ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b02fbada-b06b-44a4-a5d1-548b47882594",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# PostgreSQL connection details\n",
    "host = \"localhost\"\n",
    "dbname = \"IVF\"\n",
    "user = \"postgres\"\n",
    "password = os.getenv(\"PASSWORD\")\n",
    "\n",
    "# Create an SQLAlchemy engine\n",
    "engine = create_engine(f\"postgresql://{user}:{password}@{host}/{dbname}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a11b7907-0d73-4291-b103-b6c6ccaa03bf",
   "metadata": {},
   "source": [
    "### We begin with a data summary. ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "811bf17f-b3e9-4dcb-98ff-9966eb2dcb63",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Initialize the summary data structure\n",
    "summary_df = pd.DataFrame(columns=[\n",
    "    'patient_id',\n",
    "    'fertility_status',\n",
    "    'source',\n",
    "\n",
    "    # 2a: Variant Burden & Frequency\n",
    "    'count_variants',\n",
    "    'mean_af',\n",
    "    'count_af_over_20',\n",
    "    'shared_loci_count',\n",
    "\n",
    "    # 2b: Functional Impact\n",
    "    'pathogenic_clinvar_count',\n",
    "    'high_impact_function_count',\n",
    "    'damaging_sift_count',\n",
    "    'damaging_polyphen_count',\n",
    "    'mean_phred_qual_score',\n",
    "\n",
    "    # 2c: Gene & Pathway Features\n",
    "    'ivf_gene_variant_count',\n",
    "    'reproductive_go_variant_count',\n",
    "    'unique_genes_affected',\n",
    "    'embryo_gene_count',\n",
    "    'hormone_gene_variants',\n",
    "    'implantation_gene_variants',\n",
    "    'ovarian_reserve_variants',\n",
    "    'ivf_failure_variants',\n",
    "    'recurrent_pregnancy_loss_variants',\n",
    "    'epigenetic_variants',\n",
    "    'age_related_variants',\n",
    "    'allcock_variants',\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e94007a5-7a29-4e45-b61f-bdf5120002c1",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r6/lqyyhhmj4y5gdvd9d1nf04c40000gn/T/ipykernel_37033/3783648215.py:244: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  summary_df = pd.concat([summary_df, pd.DataFrame([results])], ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# Summary analysis\n",
    "patient_ids = [1, 2, 3, 4, 5, 6, 7]\n",
    "sources = ['germline','somatic']\n",
    "\n",
    "for pid in patient_ids:\n",
    "    \n",
    "    for source in sources:\n",
    "\n",
    "        # fertility\n",
    "        fertility_query = f\"\"\"\n",
    "            SELECT fertility \n",
    "            FROM fertility_status\n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        fertility_df = pd.read_sql(fertility_query, engine)\n",
    "        if fertility_df.empty:\n",
    "            print(f\"patient_id {pid} not found, exiting loop.\")\n",
    "            break  # or use `continue` if you want to skip this one but keep looping\n",
    "        else:\n",
    "            fertility_status = fertility_df.iloc[0]['fertility']\n",
    "    \n",
    "        # Count Variants\n",
    "        count_variants_query = f\"\"\" \n",
    "            SELECT COUNT(*) AS count \n",
    "            FROM {source} \n",
    "            WHERE patient_id = {pid} \n",
    "        \"\"\"\n",
    "        count_variants = pd.read_sql(count_variants_query, engine).iloc[0]['count']\n",
    "    \n",
    "        # Allele frequency\n",
    "        allele_freq_query = f\"\"\" \n",
    "            SELECT allele_frequency_percent \n",
    "            FROM {source} \n",
    "            WHERE patient_id = {pid} \n",
    "        \"\"\"\n",
    "        allele_freq_df = pd.read_sql(allele_freq_query, engine)\n",
    "        allele_freq_df['allele_frequency_percent'] = pd.to_numeric(allele_freq_df['allele_frequency_percent'], errors='coerce')\n",
    "        # Mean\n",
    "        mean_af = allele_freq_df['allele_frequency_percent'].mean()   \n",
    "        # Counts over threshold\n",
    "        allele_freq_df['af_gt_20'] = allele_freq_df['allele_frequency_percent'] > 20\n",
    "        count_af_over_20 = allele_freq_df['af_gt_20'].sum()\n",
    "    \n",
    "        # Shared loci count\n",
    "        baseline_loci_query = f\"\"\"\n",
    "            SELECT DISTINCT locus\n",
    "            FROM germline\n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        baseline_loci_df = pd.read_sql(baseline_loci_query, engine)\n",
    "        baseline_loci = set(baseline_loci_df['locus'])   \n",
    "        somatic_loci_query = f\"\"\"\n",
    "            SELECT DISTINCT locus\n",
    "            FROM somatic\n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        somatic_loci_df = pd.read_sql(somatic_loci_query, engine)\n",
    "        somatic_loci = set(somatic_loci_df['locus'])    \n",
    "        shared_loci = baseline_loci.intersection(somatic_loci)\n",
    "        shared_loci_count = len(shared_loci)\n",
    "    \n",
    "        # Pathogenic count  \n",
    "        clinvar_query = f\"\"\" \n",
    "            SELECT clinvar \n",
    "            FROM {source} \n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        clinvar_df = pd.read_sql(clinvar_query, engine)\n",
    "        pathogenic_clinvar_count = clinvar_df[clinvar_df['clinvar'].str.contains('pathogenic', case=False, na=False)]['clinvar'].count()\n",
    "    \n",
    "        # High impact function count\n",
    "        high_impact_terms = ['frameshift', 'stop_gained', 'splice_donor', 'splice_acceptor']\n",
    "        function_query = f\"\"\" \n",
    "            SELECT function \n",
    "            FROM {source} \n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        function_df = pd.read_sql(function_query, engine)\n",
    "        high_impact_function_count = function_df[function_df['function'].str.contains('|'.join(high_impact_terms), case=False, na=False)]['function'].count()\n",
    "    \n",
    "        # Damaging sift count\n",
    "        sift_query = f\"\"\" \n",
    "            SELECT sift \n",
    "            FROM {source}\n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        sift_df = pd.read_sql(sift_query, engine)\n",
    "        sift_df['sift'] = pd.to_numeric(sift_df['sift'], errors='coerce')\n",
    "        damaging_sift_count = sift_df[sift_df['sift'] < 0.05]['sift'].count()\n",
    "    \n",
    "        # Damaging polyphen count\n",
    "        polyphen_query = f\"\"\" \n",
    "            SELECT polyphen \n",
    "            FROM {source} \n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        polyphen_df = pd.read_sql(polyphen_query, engine)\n",
    "        polyphen_df['polyphen'] = pd.to_numeric(polyphen_df['polyphen'], errors='coerce')\n",
    "        damaging_polyphen_count = polyphen_df[polyphen_df['polyphen'] > 0.85]['polyphen'].count()\n",
    "    \n",
    "        # Mean phred qual score\n",
    "        phred_qual_score_query = f\"\"\" \n",
    "            SELECT phred_qual_score \n",
    "            FROM {source} \n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        phred_qual_score_df = pd.read_sql(phred_qual_score_query, engine)\n",
    "        phred_qual_score_df['phred_qual_score'] = pd.to_numeric(phred_qual_score_df['phred_qual_score'], errors='coerce')\n",
    "        mean_phred_qual_score = phred_qual_score_df['phred_qual_score'].mean()\n",
    "    \n",
    "        # IVF gene variant count\n",
    "        ivf_genes = ['BRCA1', 'BRCA2', 'FMR1', 'NR5A1', 'ESR1', 'FSHR', 'LHCGR', 'AMH', 'BMP15', 'GDF9', 'ZP3', 'ZP1', 'ZP2', 'NANOS3', 'NOBOX', 'FIGLA']\n",
    "        gene_query = f\"\"\" \n",
    "            SELECT gene\n",
    "            FROM {source} \n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        gene_df = pd.read_sql(gene_query, engine)\n",
    "        ivf_gene_variant_count = gene_df[gene_df['gene'].isin(ivf_genes)]['gene'].count()\n",
    "        unique_genes_affected = gene_df['gene'].nunique()\n",
    "        embryo_dev_genes = ['POU5F1', 'NANOG', 'SOX2', 'TDGF1', 'DAZL', 'MTHFR', 'PRDM9']\n",
    "        embryo_gene_count = gene_df[gene_df['gene'].isin(embryo_dev_genes)]['gene'].count()\n",
    "        hormone_genes = ['FSHR', 'LHCGR', 'ESR1', 'ESR2', 'PGR', 'AR', 'AMH', 'AMHR2']\n",
    "        hormone_gene_variants = gene_df[gene_df['gene'].isin(hormone_genes)]['gene'].count()\n",
    "        implantation_genes = ['HLA-G', 'VEGF', 'LIF', 'HOXA10', 'HOXA11', 'MUC1', 'ITGB3']\n",
    "        implantation_gene_variants = gene_df[gene_df['gene'].isin(implantation_genes)]['gene'].count()\n",
    "        ovarian_reserve_genes = ['AMH', 'AMHR2', 'FSHR', 'BMP15', 'GDF9', 'FMR1']\n",
    "        ovarian_reserve_variants = gene_df[gene_df['gene'].isin(ovarian_reserve_genes)]['gene'].count()\n",
    "        ivf_failure_genes = ['TP53', 'MTHFR', 'F5', 'F2', 'PAI1', 'ANXA5']\n",
    "        ivf_failure_variants = gene_df[gene_df['gene'].isin(ivf_failure_genes)]['gene'].count()\n",
    "        rpl_genes = ['MTHFR', 'F5', 'F2', 'ANXA5', 'PAI1', 'HLA-G', 'KIR', 'NLRP7']\n",
    "        rpl_variants = gene_df[gene_df['gene'].isin(rpl_genes)]['gene'].count()\n",
    "        epigenetic_genes = ['DNMT1', 'DNMT3A', 'DNMT3B', 'TET1', 'TET2', 'TET3', 'HDAC1', 'HDAC2']\n",
    "        epigenetic_variants = gene_df[gene_df['gene'].isin(epigenetic_genes)]['gene'].count()\n",
    "        # Assuming you have patient age data\n",
    "        age_related_genes = ['AMH', 'AMHR2', 'BMP15', 'BRCA1', 'BRCA2', 'TP53']\n",
    "        age_related_variants = gene_df[gene_df['gene'].isin(age_related_genes)]['gene'].count()\n",
    "        allcock_genes = ['CDC6','CCNA2','CCND1','CDKN2A','IL1R1','MCM2','MDM2','MKI67','PCNA','PDGFRA','PGR','SMAD3']\n",
    "        allcock_variants = gene_df[gene_df['gene'].isin(allcock_genes)]['gene'].count()\n",
    "        \n",
    "        # Reproductive go variant count\n",
    "        repro_terms = ['reproduction', 'fertilization', 'oogenesis', 'spermatogenesis', 'embryo','gonad', 'infertility', 'implantation', 'oocyte', 'sperm', 'zygote','follicle', 'blastocyst', 'luteinizing', 'FSH', 'hormone', 'gonadotropin']\n",
    "        repro_go_ids = set()\n",
    "        for go_id, term in go_dag.items():\n",
    "            name = term.name.lower()\n",
    "            definition = getattr(term, 'definition', getattr(term, 'defn', '')).lower()\n",
    "            if any(r in name or r in definition for r in repro_terms):\n",
    "                repro_go_ids.add(go_id)\n",
    "        go_query = f\"\"\" \n",
    "            SELECT go\n",
    "            FROM {source} \n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        go_df = pd.read_sql(go_query, engine)\n",
    "        hg19_go_query = f\"\"\"\n",
    "            SELECT hg19_go_1218\n",
    "            FROM {source}\n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        hg19_go_df = pd.read_sql(hg19_go_query, engine)\n",
    "        all_go_terms = pd.concat([\n",
    "            go_df['go'].dropna().astype(str),\n",
    "            hg19_go_df['hg19_go_1218'].dropna().astype(str)\n",
    "        ])\n",
    "        import re\n",
    "        go_pattern = '|'.join(map(re.escape, repro_go_ids))\n",
    "    \n",
    "        if not go_pattern:\n",
    "            reproductive_go_variant_count = 0\n",
    "        else:\n",
    "            reproductive_go_variant_count = all_go_terms[\n",
    "                all_go_terms.str.contains(go_pattern, case=False, na=False)\n",
    "            ].count()\n",
    "\n",
    "        # hg19_phylop\n",
    "        phylop_query = f\"\"\"\n",
    "            SELECT hg19_phylop_1\n",
    "            FROM {source}\n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        hg19_phylop_df = pd.read_sql(phylop_query, engine)\n",
    "        hg19_phylop_df['hg19_phylop_1'] = pd.to_numeric(hg19_phylop_df['hg19_phylop_1'], errors='coerce')\n",
    "        mean_hg19_phylop = hg19_phylop_df['hg19_phylop_1'].mean()\n",
    "        hg19_phylop_conserved_count = hg19_phylop_df[hg19_phylop_df['hg19_phylop_1'] > 2.0]['hg19_phylop_1'].count()\n",
    "\n",
    "        # hg19_cosmic\n",
    "        hg19_cosmic_query = f\"\"\"\n",
    "            SELECT hg19_cosmic_67\n",
    "            FROM {source}\n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        hg19_cosmic_df = pd.read_sql(hg19_cosmic_query, engine)\n",
    "        hg19_cosmic_count = hg19_cosmic_df['hg19_cosmic_67'].notna().sum()\n",
    "\n",
    "        # hg19_esp6500\n",
    "        esp_query = f\"\"\"\n",
    "            SELECT hg19_esp6500_1\n",
    "            FROM {source}\n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        esp_df = pd.read_sql(esp_query, engine)\n",
    "        esp_df['hg19_esp6500_1'] = pd.to_numeric(esp_df['hg19_esp6500_1'], errors='coerce')\n",
    "        hg19_esp_mean = esp_df['hg19_esp6500_1'].mean()\n",
    "        hg19_esp_rare_variants = esp_df[esp_df['hg19_esp6500_1'] < 0.01]['hg19_esp6500_1'].count()\n",
    "\n",
    "        # hg19_pfam\n",
    "        pfam_query = f\"\"\"\n",
    "            SELECT hg19_pfam_26\n",
    "            FROM {source}\n",
    "            WHERE patient_id = {pid}\n",
    "        \"\"\"\n",
    "        pfam_df = pd.read_sql(pfam_query, engine)\n",
    "        hg19_pfam_count = pfam_df['hg19_pfam_26'].notna().sum()\n",
    "    \n",
    "        # Append results\n",
    "        results = {\n",
    "            'patient_id': pid,\n",
    "            'fertility_status': fertility_status,\n",
    "            'source': source,\n",
    "            'count_variants': count_variants,\n",
    "            'mean_af': mean_af,\n",
    "            'count_af_over_20': count_af_over_20,\n",
    "            'shared_loci_count': shared_loci_count,\n",
    "            'pathogenic_clinvar_count': pathogenic_clinvar_count,\n",
    "            'high_impact_function_count': high_impact_function_count,\n",
    "            'damaging_sift_count': damaging_sift_count,\n",
    "            'damaging_polyphen_count': damaging_polyphen_count,\n",
    "            'mean_phred_qual_score': mean_phred_qual_score,\n",
    "            'ivf_gene_variant_count': ivf_gene_variant_count,\n",
    "            'unique_genes_affected': unique_genes_affected,\n",
    "            'embryo_gene_count': embryo_gene_count,\n",
    "            'hormone_gene_variants': hormone_gene_variants,\n",
    "            'reproductive_go_variant_count': reproductive_go_variant_count,\n",
    "            'implantation_gene_variants': implantation_gene_variants,\n",
    "            'ovarian_reserve_variants': ovarian_reserve_variants,\n",
    "            'ivf_failure_variants': ivf_failure_variants,\n",
    "            'recurrent_pregnancy_loss_variants': rpl_variants,\n",
    "            'epigenetic_variants': epigenetic_variants,\n",
    "            'age_related_variants': age_related_variants,\n",
    "            'allcock_variants': allcock_variants,\n",
    "        }\n",
    "    \n",
    "        # Add row to summary\n",
    "        summary_df = pd.concat([summary_df, pd.DataFrame([results])], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6e896719-b2bc-4088-9816-651811027c95",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>patient_id</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fertility_status</th>\n",
       "      <td>primary infertility</td>\n",
       "      <td>primary infertility</td>\n",
       "      <td>primary infertility</td>\n",
       "      <td>primary infertility</td>\n",
       "      <td>fertile</td>\n",
       "      <td>fertile</td>\n",
       "      <td>primary infertility</td>\n",
       "      <td>primary infertility</td>\n",
       "      <td>primary infertility</td>\n",
       "      <td>primary infertility</td>\n",
       "      <td>primary infertility</td>\n",
       "      <td>primary infertility</td>\n",
       "      <td>fertile</td>\n",
       "      <td>fertile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>source</th>\n",
       "      <td>germline</td>\n",
       "      <td>somatic</td>\n",
       "      <td>germline</td>\n",
       "      <td>somatic</td>\n",
       "      <td>germline</td>\n",
       "      <td>somatic</td>\n",
       "      <td>germline</td>\n",
       "      <td>somatic</td>\n",
       "      <td>germline</td>\n",
       "      <td>somatic</td>\n",
       "      <td>germline</td>\n",
       "      <td>somatic</td>\n",
       "      <td>germline</td>\n",
       "      <td>somatic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count_variants</th>\n",
       "      <td>50263</td>\n",
       "      <td>46330</td>\n",
       "      <td>52099</td>\n",
       "      <td>53132</td>\n",
       "      <td>52125</td>\n",
       "      <td>55451</td>\n",
       "      <td>49572</td>\n",
       "      <td>43913</td>\n",
       "      <td>52394</td>\n",
       "      <td>52362</td>\n",
       "      <td>52936</td>\n",
       "      <td>59137</td>\n",
       "      <td>54406</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean_af</th>\n",
       "      <td>68.7805</td>\n",
       "      <td>65.120089</td>\n",
       "      <td>68.50133</td>\n",
       "      <td>65.704918</td>\n",
       "      <td>68.314328</td>\n",
       "      <td>63.766697</td>\n",
       "      <td>69.471021</td>\n",
       "      <td>65.754442</td>\n",
       "      <td>68.789142</td>\n",
       "      <td>66.012886</td>\n",
       "      <td>68.227191</td>\n",
       "      <td>63.898222</td>\n",
       "      <td>67.610661</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count_af_over_20</th>\n",
       "      <td>49373</td>\n",
       "      <td>40293</td>\n",
       "      <td>51044</td>\n",
       "      <td>45719</td>\n",
       "      <td>51155</td>\n",
       "      <td>46436</td>\n",
       "      <td>48774</td>\n",
       "      <td>39311</td>\n",
       "      <td>51353</td>\n",
       "      <td>45927</td>\n",
       "      <td>51653</td>\n",
       "      <td>47318</td>\n",
       "      <td>53450</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>shared_loci_count</th>\n",
       "      <td>40556</td>\n",
       "      <td>40556</td>\n",
       "      <td>46306</td>\n",
       "      <td>46306</td>\n",
       "      <td>46686</td>\n",
       "      <td>46686</td>\n",
       "      <td>39528</td>\n",
       "      <td>39528</td>\n",
       "      <td>46464</td>\n",
       "      <td>46464</td>\n",
       "      <td>47845</td>\n",
       "      <td>47845</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pathogenic_clinvar_count</th>\n",
       "      <td>237</td>\n",
       "      <td>201</td>\n",
       "      <td>195</td>\n",
       "      <td>186</td>\n",
       "      <td>188</td>\n",
       "      <td>173</td>\n",
       "      <td>190</td>\n",
       "      <td>161</td>\n",
       "      <td>229</td>\n",
       "      <td>216</td>\n",
       "      <td>210</td>\n",
       "      <td>194</td>\n",
       "      <td>208</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>high_impact_function_count</th>\n",
       "      <td>420</td>\n",
       "      <td>1025</td>\n",
       "      <td>421</td>\n",
       "      <td>1012</td>\n",
       "      <td>403</td>\n",
       "      <td>1323</td>\n",
       "      <td>381</td>\n",
       "      <td>1073</td>\n",
       "      <td>447</td>\n",
       "      <td>1008</td>\n",
       "      <td>420</td>\n",
       "      <td>1300</td>\n",
       "      <td>373</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>damaging_sift_count</th>\n",
       "      <td>1329</td>\n",
       "      <td>1291</td>\n",
       "      <td>1356</td>\n",
       "      <td>1370</td>\n",
       "      <td>1356</td>\n",
       "      <td>1524</td>\n",
       "      <td>1304</td>\n",
       "      <td>1183</td>\n",
       "      <td>1400</td>\n",
       "      <td>1385</td>\n",
       "      <td>1380</td>\n",
       "      <td>1521</td>\n",
       "      <td>1389</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>damaging_polyphen_count</th>\n",
       "      <td>861</td>\n",
       "      <td>849</td>\n",
       "      <td>897</td>\n",
       "      <td>912</td>\n",
       "      <td>919</td>\n",
       "      <td>1066</td>\n",
       "      <td>883</td>\n",
       "      <td>819</td>\n",
       "      <td>941</td>\n",
       "      <td>941</td>\n",
       "      <td>890</td>\n",
       "      <td>1031</td>\n",
       "      <td>937</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean_phred_qual_score</th>\n",
       "      <td>372.097499</td>\n",
       "      <td>714.718106</td>\n",
       "      <td>692.487963</td>\n",
       "      <td>1287.499363</td>\n",
       "      <td>524.794417</td>\n",
       "      <td>917.425845</td>\n",
       "      <td>307.051277</td>\n",
       "      <td>593.764199</td>\n",
       "      <td>635.181634</td>\n",
       "      <td>1180.992057</td>\n",
       "      <td>651.606178</td>\n",
       "      <td>1125.70447</td>\n",
       "      <td>707.279515</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ivf_gene_variant_count</th>\n",
       "      <td>47</td>\n",
       "      <td>42</td>\n",
       "      <td>53</td>\n",
       "      <td>57</td>\n",
       "      <td>63</td>\n",
       "      <td>62</td>\n",
       "      <td>35</td>\n",
       "      <td>34</td>\n",
       "      <td>59</td>\n",
       "      <td>61</td>\n",
       "      <td>42</td>\n",
       "      <td>52</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reproductive_go_variant_count</th>\n",
       "      <td>4042</td>\n",
       "      <td>3554</td>\n",
       "      <td>4199</td>\n",
       "      <td>4091</td>\n",
       "      <td>4154</td>\n",
       "      <td>4190</td>\n",
       "      <td>4043</td>\n",
       "      <td>3576</td>\n",
       "      <td>4238</td>\n",
       "      <td>4087</td>\n",
       "      <td>4315</td>\n",
       "      <td>4411</td>\n",
       "      <td>4466</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique_genes_affected</th>\n",
       "      <td>14176</td>\n",
       "      <td>14852</td>\n",
       "      <td>14316</td>\n",
       "      <td>15743</td>\n",
       "      <td>14214</td>\n",
       "      <td>16103</td>\n",
       "      <td>14032</td>\n",
       "      <td>13738</td>\n",
       "      <td>14299</td>\n",
       "      <td>15168</td>\n",
       "      <td>14608</td>\n",
       "      <td>17797</td>\n",
       "      <td>14553</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>embryo_gene_count</th>\n",
       "      <td>18</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>16</td>\n",
       "      <td>11</td>\n",
       "      <td>15</td>\n",
       "      <td>13</td>\n",
       "      <td>16</td>\n",
       "      <td>17</td>\n",
       "      <td>21</td>\n",
       "      <td>18</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hormone_gene_variants</th>\n",
       "      <td>16</td>\n",
       "      <td>17</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>19</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>18</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>implantation_gene_variants</th>\n",
       "      <td>28</td>\n",
       "      <td>24</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>31</td>\n",
       "      <td>27</td>\n",
       "      <td>20</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ovarian_reserve_variants</th>\n",
       "      <td>13</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ivf_failure_variants</th>\n",
       "      <td>31</td>\n",
       "      <td>32</td>\n",
       "      <td>30</td>\n",
       "      <td>27</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>22</td>\n",
       "      <td>18</td>\n",
       "      <td>20</td>\n",
       "      <td>18</td>\n",
       "      <td>21</td>\n",
       "      <td>20</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>recurrent_pregnancy_loss_variants</th>\n",
       "      <td>56</td>\n",
       "      <td>50</td>\n",
       "      <td>47</td>\n",
       "      <td>45</td>\n",
       "      <td>46</td>\n",
       "      <td>45</td>\n",
       "      <td>53</td>\n",
       "      <td>45</td>\n",
       "      <td>29</td>\n",
       "      <td>28</td>\n",
       "      <td>50</td>\n",
       "      <td>53</td>\n",
       "      <td>47</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>epigenetic_variants</th>\n",
       "      <td>17</td>\n",
       "      <td>18</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>17</td>\n",
       "      <td>21</td>\n",
       "      <td>17</td>\n",
       "      <td>14</td>\n",
       "      <td>26</td>\n",
       "      <td>27</td>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>age_related_variants</th>\n",
       "      <td>25</td>\n",
       "      <td>23</td>\n",
       "      <td>27</td>\n",
       "      <td>28</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>23</td>\n",
       "      <td>25</td>\n",
       "      <td>12</td>\n",
       "      <td>19</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allcock_variants</th>\n",
       "      <td>28</td>\n",
       "      <td>26</td>\n",
       "      <td>37</td>\n",
       "      <td>35</td>\n",
       "      <td>39</td>\n",
       "      <td>45</td>\n",
       "      <td>42</td>\n",
       "      <td>39</td>\n",
       "      <td>43</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0                    1   \\\n",
       "patient_id                                           1                    1   \n",
       "fertility_status                   primary infertility  primary infertility   \n",
       "source                                        germline              somatic   \n",
       "count_variants                                   50263                46330   \n",
       "mean_af                                        68.7805            65.120089   \n",
       "count_af_over_20                                 49373                40293   \n",
       "shared_loci_count                                40556                40556   \n",
       "pathogenic_clinvar_count                           237                  201   \n",
       "high_impact_function_count                         420                 1025   \n",
       "damaging_sift_count                               1329                 1291   \n",
       "damaging_polyphen_count                            861                  849   \n",
       "mean_phred_qual_score                       372.097499           714.718106   \n",
       "ivf_gene_variant_count                              47                   42   \n",
       "reproductive_go_variant_count                     4042                 3554   \n",
       "unique_genes_affected                            14176                14852   \n",
       "embryo_gene_count                                   18                   14   \n",
       "hormone_gene_variants                               16                   17   \n",
       "implantation_gene_variants                          28                   24   \n",
       "ovarian_reserve_variants                            13                   10   \n",
       "ivf_failure_variants                                31                   32   \n",
       "recurrent_pregnancy_loss_variants                   56                   50   \n",
       "epigenetic_variants                                 17                   18   \n",
       "age_related_variants                                25                   23   \n",
       "allcock_variants                                    28                   26   \n",
       "\n",
       "                                                    2                    3   \\\n",
       "patient_id                                           2                    2   \n",
       "fertility_status                   primary infertility  primary infertility   \n",
       "source                                        germline              somatic   \n",
       "count_variants                                   52099                53132   \n",
       "mean_af                                       68.50133            65.704918   \n",
       "count_af_over_20                                 51044                45719   \n",
       "shared_loci_count                                46306                46306   \n",
       "pathogenic_clinvar_count                           195                  186   \n",
       "high_impact_function_count                         421                 1012   \n",
       "damaging_sift_count                               1356                 1370   \n",
       "damaging_polyphen_count                            897                  912   \n",
       "mean_phred_qual_score                       692.487963          1287.499363   \n",
       "ivf_gene_variant_count                              53                   57   \n",
       "reproductive_go_variant_count                     4199                 4091   \n",
       "unique_genes_affected                            14316                15743   \n",
       "embryo_gene_count                                   14                   13   \n",
       "hormone_gene_variants                               22                   22   \n",
       "implantation_gene_variants                          22                   22   \n",
       "ovarian_reserve_variants                             7                    8   \n",
       "ivf_failure_variants                                30                   27   \n",
       "recurrent_pregnancy_loss_variants                   47                   45   \n",
       "epigenetic_variants                                 19                   19   \n",
       "age_related_variants                                27                   28   \n",
       "allcock_variants                                    37                   35   \n",
       "\n",
       "                                           4           5   \\\n",
       "patient_id                                  3           3   \n",
       "fertility_status                      fertile     fertile   \n",
       "source                               germline     somatic   \n",
       "count_variants                          52125       55451   \n",
       "mean_af                             68.314328   63.766697   \n",
       "count_af_over_20                        51155       46436   \n",
       "shared_loci_count                       46686       46686   \n",
       "pathogenic_clinvar_count                  188         173   \n",
       "high_impact_function_count                403        1323   \n",
       "damaging_sift_count                      1356        1524   \n",
       "damaging_polyphen_count                   919        1066   \n",
       "mean_phred_qual_score              524.794417  917.425845   \n",
       "ivf_gene_variant_count                     63          62   \n",
       "reproductive_go_variant_count            4154        4190   \n",
       "unique_genes_affected                   14214       16103   \n",
       "embryo_gene_count                          16          11   \n",
       "hormone_gene_variants                      19          19   \n",
       "implantation_gene_variants                 31          27   \n",
       "ovarian_reserve_variants                   12          12   \n",
       "ivf_failure_variants                       13          13   \n",
       "recurrent_pregnancy_loss_variants          46          45   \n",
       "epigenetic_variants                        17          21   \n",
       "age_related_variants                       31          31   \n",
       "allcock_variants                           39          45   \n",
       "\n",
       "                                                    6                    7   \\\n",
       "patient_id                                           4                    4   \n",
       "fertility_status                   primary infertility  primary infertility   \n",
       "source                                        germline              somatic   \n",
       "count_variants                                   49572                43913   \n",
       "mean_af                                      69.471021            65.754442   \n",
       "count_af_over_20                                 48774                39311   \n",
       "shared_loci_count                                39528                39528   \n",
       "pathogenic_clinvar_count                           190                  161   \n",
       "high_impact_function_count                         381                 1073   \n",
       "damaging_sift_count                               1304                 1183   \n",
       "damaging_polyphen_count                            883                  819   \n",
       "mean_phred_qual_score                       307.051277           593.764199   \n",
       "ivf_gene_variant_count                              35                   34   \n",
       "reproductive_go_variant_count                     4043                 3576   \n",
       "unique_genes_affected                            14032                13738   \n",
       "embryo_gene_count                                   15                   13   \n",
       "hormone_gene_variants                               18                   19   \n",
       "implantation_gene_variants                          20                   16   \n",
       "ovarian_reserve_variants                             9                    7   \n",
       "ivf_failure_variants                                22                   18   \n",
       "recurrent_pregnancy_loss_variants                   53                   45   \n",
       "epigenetic_variants                                 17                   14   \n",
       "age_related_variants                                15                   14   \n",
       "allcock_variants                                    42                   39   \n",
       "\n",
       "                                                    8                    9   \\\n",
       "patient_id                                           5                    5   \n",
       "fertility_status                   primary infertility  primary infertility   \n",
       "source                                        germline              somatic   \n",
       "count_variants                                   52394                52362   \n",
       "mean_af                                      68.789142            66.012886   \n",
       "count_af_over_20                                 51353                45927   \n",
       "shared_loci_count                                46464                46464   \n",
       "pathogenic_clinvar_count                           229                  216   \n",
       "high_impact_function_count                         447                 1008   \n",
       "damaging_sift_count                               1400                 1385   \n",
       "damaging_polyphen_count                            941                  941   \n",
       "mean_phred_qual_score                       635.181634          1180.992057   \n",
       "ivf_gene_variant_count                              59                   61   \n",
       "reproductive_go_variant_count                     4238                 4087   \n",
       "unique_genes_affected                            14299                15168   \n",
       "embryo_gene_count                                   16                   17   \n",
       "hormone_gene_variants                               15                   16   \n",
       "implantation_gene_variants                           2                    2   \n",
       "ovarian_reserve_variants                             7                    7   \n",
       "ivf_failure_variants                                20                   18   \n",
       "recurrent_pregnancy_loss_variants                   29                   28   \n",
       "epigenetic_variants                                 26                   27   \n",
       "age_related_variants                                23                   25   \n",
       "allcock_variants                                    43                   39   \n",
       "\n",
       "                                                    10                   11  \\\n",
       "patient_id                                           6                    6   \n",
       "fertility_status                   primary infertility  primary infertility   \n",
       "source                                        germline              somatic   \n",
       "count_variants                                   52936                59137   \n",
       "mean_af                                      68.227191            63.898222   \n",
       "count_af_over_20                                 51653                47318   \n",
       "shared_loci_count                                47845                47845   \n",
       "pathogenic_clinvar_count                           210                  194   \n",
       "high_impact_function_count                         420                 1300   \n",
       "damaging_sift_count                               1380                 1521   \n",
       "damaging_polyphen_count                            890                 1031   \n",
       "mean_phred_qual_score                       651.606178           1125.70447   \n",
       "ivf_gene_variant_count                              42                   52   \n",
       "reproductive_go_variant_count                     4315                 4411   \n",
       "unique_genes_affected                            14608                17797   \n",
       "embryo_gene_count                                   21                   18   \n",
       "hormone_gene_variants                               16                   18   \n",
       "implantation_gene_variants                          28                   27   \n",
       "ovarian_reserve_variants                             9                   10   \n",
       "ivf_failure_variants                                21                   20   \n",
       "recurrent_pregnancy_loss_variants                   50                   53   \n",
       "epigenetic_variants                                 13                   18   \n",
       "age_related_variants                                12                   19   \n",
       "allcock_variants                                    39                   44   \n",
       "\n",
       "                                           12       13  \n",
       "patient_id                                  7        7  \n",
       "fertility_status                      fertile  fertile  \n",
       "source                               germline  somatic  \n",
       "count_variants                          54406        0  \n",
       "mean_af                             67.610661      NaN  \n",
       "count_af_over_20                        53450        0  \n",
       "shared_loci_count                           0        0  \n",
       "pathogenic_clinvar_count                  208        0  \n",
       "high_impact_function_count                373        0  \n",
       "damaging_sift_count                      1389        0  \n",
       "damaging_polyphen_count                   937        0  \n",
       "mean_phred_qual_score              707.279515      NaN  \n",
       "ivf_gene_variant_count                     60        0  \n",
       "reproductive_go_variant_count            4466        0  \n",
       "unique_genes_affected                   14553        0  \n",
       "embryo_gene_count                          19        0  \n",
       "hormone_gene_variants                      19        0  \n",
       "implantation_gene_variants                 33        0  \n",
       "ovarian_reserve_variants                    7        0  \n",
       "ivf_failure_variants                       22        0  \n",
       "recurrent_pregnancy_loss_variants          47        0  \n",
       "epigenetic_variants                        22        0  \n",
       "age_related_variants                       27        0  \n",
       "allcock_variants                           44        0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display\n",
    "display(summary_df.T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0e3cb39-bcd0-4ea5-9f46-c9b743f4ff22",
   "metadata": {},
   "source": [
    "# Initial Findings #\n",
    "## Fertility Status Correlation ##\n",
    "\n",
    "Variant Counts: Patients with primary infertility generally have a high number of genomic variants.\n",
    "\n",
    "Shared Loci Count: Primary infertility shows similarities in shared loci across patients.\n",
    "\n",
    "Mean AF & Pathogenic Variants: No significant difference in allele frequencies or pathogenic variants between fertile and infertile groups.\n",
    "\n",
    "## Genomic Insights ##\n",
    "\n",
    "Source of Variants:\n",
    "Germline and somatic classifications are consistent across patients, with no apparent impact on fertility status.\n",
    "\n",
    "Functional Impact:\n",
    "High-impact function counts are higher in patients with primary infertility.\n",
    "Damaging SIFT and PolyPhen counts also tend to be higher among these patients.\n",
    "\n",
    "## Reproductive-specific Variants ##\n",
    "\n",
    "IVF and Reproductive Variants:\n",
    "IVF and other reproductive gene variants don't significantly distinguish fertile from infertile, but IVF gene variant counts are lower in fertile individuals.\n",
    "\n",
    "Embryo & Hormone Genes:\n",
    "Minimal variations in these counts across the groups, suggesting no strong correlation with fertility status.\n",
    "\n",
    "## Other Observations ##\n",
    "\n",
    "Age-related and Recurrent Pregnancy Loss Variants:\n",
    "Age-related variants are slightly more frequent among fertile individuals.\n",
    "Recurrent pregnancy loss variants are common, but not exclusively in infertile patients.\n",
    "\n",
    "## Conclusion ##\n",
    "\n",
    "Patients with primary infertility tend to have high variant counts and more high-impact functional variants. However, the specific role of these findings in fertility remains unclear without a richer data set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85028eef-92cd-4bbb-8d67-79db32c7c2ab",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
